unit_tests/ex_suffix_expander.cc

```diff
@@ -9,13 +9,177 @@
+    } else if (unit == TB) {
+      return static_cast<uintmax_t>(1) << 40;
+    } else if (unit == PB) {
+      return static_cast<uintmax_t>(1) << 50;
+    } else {
+      // TODO: error handling
+      return static_cast<uintmax_t>(0);
+    }
+  }
+
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "b") || iequal(unit, "byte")) {
+      return B;
+    } else if (iequal(unit, "k") || iequal(unit, "kb")) {
+      return KB;
+    } else if (iequal(unit, "m") || iequal(unit, "mb")) {
+      return MB;
+    } else if (iequal(unit, "g") || iequal(unit, "gb")) {
+      return GB;
+    } else if (iequal(unit, "t") || iequal(unit, "tb")) {
+      return TB;
+    } else if (iequal(unit, "p") || iequal(unit, "pb")) {
+      return PB;
+    } else {
+      // TODO: error handling
+      return B;
+    }
+  }
+};
+
+struct Duration {
+  static constexpr TextView SECOND = "second";
+  static constexpr TextView MINUTE = "minute";
+  static constexpr TextView HOUR   = "hour";
+  static constexpr TextView DAY    = "day";
+  static constexpr TextView WEEK   = "week";
+
+  static constexpr TextView baseUnit = "second";
+
+  static uintmax_t convertToBase(TextView const unit) {
+    if (unit == SECOND) {
+      return static_cast<uintmax_t>(1);
+    } else if (unit == MINUTE) {
+      return static_cast<uintmax_t>(60);
+    } else if (unit == HOUR) {
+      return static_cast<uintmax_t>(3'600);
+    } else if (unit == DAY) {
+      return static_cast<uintmax_t>(86'400);
+    } else if (unit == WEEK) {
+      return static_cast<uintmax_t>(604'800);
+    } else {
+      // TODO: error handling
+      return 0;
+    }
+  }
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "s") || iequal(unit, "sec") || iequal(unit, "second")) {
+      return SECOND;
+    } else if (iequal(unit, "m") || iequal(unit, "min") || iequal(unit, "minute")) {
+      return MINUTE;
+    } else if (iequal(unit, "h") || iequal(unit, "hour")) {
+      return HOUR;
+    } else if (iequal(unit, "d") || iequal(unit, "day")) {
+      return DAY;
+    } else if (iequal(unit, "w") || iequal(unit, "week")) {
+      return WEEK;
+    } else {
+      // TODO: error handling
+      return SECOND;
+    }
+  }
+};
+
+} // namespace metric
+
+// (potentially) modifies the pass-in view. returns the unit
+template <typename Metric> inline TextView rExtractUnit(TextView &src) noexcept {
+  src.rtrim_if(&isspace);
+  if (isdigit(src.back())) {
+    // no unit specification, defaulting to the smallest unit
+    return Metric::baseUnit;
+  } else {
+    // take suffix from right to left -- from the first non-space char
+    // (inclusive) to the first digit (exclusive)
+    // can't call split/take_suffix_if because we don't want to discard any char
+    // can't call suffix_if because it can't deal with the edge case that
+    // [0, pos] is a multiplier (string)
+    auto const pos  = src.rfind_if([](char const c) { return isdigit(c) || c == ' '; });
```

Comment:
What is the advantage of declaring the argument `const`?

Reply:
ah, just a safety precaution that I just want to read c...? to prevent myself from doing things like `isdigit(++c)`..? yea prolly meaningless in such a short lambda.